marshmallow_jsonschema/base.py

```diff
@@ -250,20 +257,68 @@ def _from_union_schema(
             ]
         }
 
-    def _get_python_type(self, field):
+    def _get_python_type(self, field: fields.Field) -> typing.Type:
         """Get python type based on field subclass"""
-        for map_class, pytype in MARSHMALLOW_TO_PY_TYPES_PAIRS:
-            if issubclass(field.__class__, map_class):
-                return pytype
+        if PYTYPE_KEY in field.metadata:
+            return field.metadata[PYTYPE_KEY]
+        else:
+            for map_class, pytype in MARSHMALLOW_TO_PY_TYPES_PAIRS:
+                if issubclass(field.__class__, map_class):
+                    return pytype
 
         raise UnsupportedValueError("unsupported field type %s" % field)
 
+    def _from_custom_field_type(self, obj, field: fields.Field, type_mapping: dict[str, typing.Any]) -> typing.Dict[str, typing.Any]:
+        """Get schema definition for a custom field."""
+        json_schema = type_mapping
+
+        json_schema["title"] = field.attribute or field.name or ""
+
+        if field.dump_only:
+            json_schema["readOnly"] = True
+
+        if field.default is not missing and not callable(field.default):
+            json_schema["default"] = field.default
+
+        if ALLOW_ENUMS and isinstance(field, EnumField):
+            json_schema["enum"] = self._get_enum_values(field)
+
+        if field.allow_none:
+            previous_type = json_schema["type"]
+            json_schema["type"] = [previous_type, "null"]
+
+        # NOTE: doubled up to maintain backwards compatibility
+        metadata = field.metadata.get("metadata", {})
+        metadata.update(field.metadata)
+
+        for md_key, md_val in metadata.items():
+            if md_key in ("metadata", "name", PYTYPE_KEY):
+                continue
+            json_schema[md_key] = md_val
+
+        if "array" in json_schema["type"]:
+            json_schema["items"] = (
+                self._get_schema_for_field(obj, field.inner)
+                if isinstance(field, fields.List)
+                else {}
+            )
+
+        if "object" in json_schema["type"]:
+            json_schema["additionalProperties"] = (
+                self._get_schema_for_field(obj, field.value_field)
+                if isinstance(field, fields.Dict) and field.value_field
+                else {}
+            )
+        return json_schema
+
     def _get_schema_for_field(self, obj, field):
         """Get schema and validators for field."""
-        if hasattr(field, "_jsonschema_type_mapping"):
```

Comment:
Do we need a helper function to handle where / how some of these things are stored in metadata?